compiler/qsc/src/interpret.rs

```diff
@@ -345,17 +348,50 @@ impl Interpreter {
     /// Performs QIR codegen using the given entry expression on a new instance of the environment
     /// and simulator but using the current compilation.
     pub fn qirgen(&mut self, expr: &str) -> std::result::Result<String, Vec<Error>> {
-        if self.capabilities != RuntimeCapabilityFlags::empty() {
+        if self.capabilities == RuntimeCapabilityFlags::all() {
             return Err(vec![Error::UnsupportedRuntimeCapabilities]);
         }
+        if self.capabilities == RuntimeCapabilityFlags::empty() {
+            let mut sim = BaseProfSim::new();
+            let mut stdout = std::io::sink();
+            let mut out = GenericReceiver::new(&mut stdout);
 
-        let mut sim = BaseProfSim::new();
-        let mut stdout = std::io::sink();
-        let mut out = GenericReceiver::new(&mut stdout);
+            let val = self.run_with_sim(&mut sim, &mut out, expr)??;
 
-        let val = self.run_with_sim(&mut sim, &mut out, expr)??;
+            Ok(sim.finish(&val))
+        } else {
+            // compile the expression which will update the
+            // entry expression in the FIR store.
+            let graph = self.compile_entry_expr(expr)?;
+            // The above line updated the entry expr in the FIR store, but we need to update
+            // the user package exec graph.
+            let package = self.fir_store.get_mut(self.package);
+            package.entry_exec_graph = graph.into();
+
+            // normal compilation has already been done
+            // analyze the capabilities before generate QIR
+            // This was already done as part of lowering, but we don't have the results here.
+            let caps_results = PassContext::run_fir_passes_on_fir(
+                &self.fir_store,
+                self.package,
+                self.capabilities,
+            );
 
-        Ok(sim.finish(&val))
+            let Ok(compute_properties) = caps_results else {
+                // This should never happen, as the program should be checked for errors before trying to
+                // generate code for it. But just in case, simply report the failure.
+                return Err(vec![Error::UnsupportedRuntimeCapabilities]);
```

Comment:
If this should never happen, should the program panic here instead of returning an error?